src/DocBlock/TypeExpression.php

```diff
@@ -428,7 +440,50 @@ private function parse(): void
                 \assert(\strlen($this->value) === $index);
 
                 $this->isUnionType = true;
-                $this->typesGlue = $typesGlue;
+
+                if (1 === \count($seenGlues)) {
+                    $this->typesGlue = array_key_first($seenGlues);
+
+                    foreach ($innerValues as $innerValue) {
+                        $this->innerTypeExpressions[] = [
+                            'start_index' => $innerValue['start_index'],
+                            'expression' => $this->inner($innerValue['value']),
+                        ];
+                    }
+                } else {
+                    $glue = null;
+                    foreach (['|', '&'] as $possibleGlue) {
+                        if ($seenGlues[$possibleGlue] ?? false) {
+                            $glue = $possibleGlue;
+
+                            break;
+                        }
+                    }
+                    \assert(null !== $glue);
+
+                    $this->typesGlue = $glue;
+
+                    for ($i = 0; $i < \count($innerValues); ++$i) {
+                        $innerStartIndex = $innerValues[$i]['start_index'];
+                        $innerValue = '';
+                        while (true) {
+                            $innerValue .= $innerValues[$i]['value'];
+
+                            if (($innerValues[$i]['next_glue'] ?? $glue) === $glue) {
+                                break;
+                            }
+
+                            $innerValue .= $innerValues[$i]['next_glue_raw'];
+
+                            ++$i;
+                        }
+
+                        $this->innerTypeExpressions[] = [
+                            'start_index' => $innerStartIndex,
+                            'expression' => $this->inner($innerValue),
+                        ];
+                    }
```

Comment:
I am not sure if I follow what's going on here. Mind providing some explanation?

Reply:
We parse `|`, `&` without priority and then push expressions glued by `&` into expressions glued with `|`. It is very optimized and actually quite short and readable code. I had to rewrite this once, see commit history.